First round migrating to a new FILE struct

elbing opened the pull request:
All comes from musl lib.
No suicides, but not all working. Still is needed
migrate rest of sources using old file struct. For now migrated all
which were using char *flags.

Printf family works at half: prints empty some formatted data.
Migration of putc/getc is done and ok :-). No more IO_putc/getc and
abandoning IO_stream.

New stdin/out/err defines and sources. And many other things.

*** fprintf suicides at: ***
if (strchr(mode, '+')) flags = O_RDWR; <-- !!! flags!
in __fmodeflags.

testwchar doesn't, it prints a puts() for debugging control and an empty string.
But, hey, this is the first round. I'm migrating almost all ape stdio.

Next step: f-> char *buf to f-> unsigned char* buf an rest of the old FILE
struct removed and migrated.

Change-Id: I9c1b48266f240164a27cb1378f1b3a373dc0e907

elbing commented:
Changed commit message:

First round migrating to a new FILE struct
All comes from musl lib.
No suicides, but not all working.

Printf family works at half: prints only the first time they're called.
No more IO_putc/getc and
Abandoned _IO_stream.

New stdin/out/err defines and sources. And many other things.

fprintf suicides at:
vfprintf.c:672

fwrite suicides at:
fwrite.c:8

putc suicides at some point in __overflow()
Look at its define in ap/internal/stdio_impl.h

testwchar doesn't suicide, it prints a puts() for debugging control and an empty string.

No more old FILE struct nor iolib.h internal header.

Inherited remaining files:

_dtoa.c
exit.c
_fconv.c
ftoa.c
pow10.c
rename.c
strerror.c
strtod.c
tmpfile.c
tmpnam.c

Adapted to work without iolib.h

rminnich commented on include/bsd.h:
```diff
@@ -24,7 +24,7 @@ extern "C" {
 #ifndef _SIZE_T
 #define _SIZE_T
-typedef unsigned long long size_t;
+typedef unsigned long size_t;
```
yuck.

rminnich commented on include/stddef.h:
```diff
@@ -32,7 +32,7 @@
 #ifndef _SIZE_T
 #define _SIZE_T
-typedef long long size_t;
+typedef unsigned long size_t;
```
so depending on order of inclusion you get different types for size_t? yuck.

elbing replied:
No, those both means only one thing: size_t was wrong. It must be unsigned long (64) and ssize_t signed long. stddef.h is temporary, in fact it's generated by compiler... when we have native one.

rminnich commented on include/stdio.h:
```diff
@@ -41,19 +42,17 @@ typedef struct FILE FILE;
 struct FILE {
 	int fd;		/* UNIX file pointer */
-	char flags;	/* bits for must free buffer on close, line-buffered */
+	char p9flags;	/* bits for must free buffer on close, line-buffered */
```
char is signed. uint8_t?

elbing replied:
huh? this is old, now FILE struct is not in stdio, just a typedef _IO_FILE FILE and moved to internal stdio_impl.h. And that part is gone. What version are you looking?

rminnich commented on include/stdio.h:
```diff
@@ -41,19 +42,17 @@ typedef struct FILE FILE;
 	size_t bufl;	/* actual length of buffer */
 	char unbuf[1];	/* tiny buffer for unbuffered io (used for ungetc?) */
 #ifdef __C99
+	unsigned flags;
```
unsigned what? int? what?

elbing replied:
The same, all that part is gone from stdio.h, no more ifdef C99

rminnich commented on include/stdio.h:
```diff
@@ -107,11 +106,19 @@ typedef long long fpos_t;
+extern FILE *const stdout;
+extern FILE *const stderr;
+
+#define stdin  (stdin)
```
this is really weird. Why?

elbing replied:
POSIX requirement, inherited from musl. std* files must be macros.

The C standards C89/C99 (the last one I even looked up right now) say 
those identifiers are macros - according to stdio.h from glibc, which 
does the following:

   /* Standard streams.  */
   extern struct _IO_FILE *stdin;          /* Standard input stream.  */
[...]
   /* C89/C99 say they're macros.  Make them happy.  */
   #define stdin stdin
[...]

rminnich commented on lib/ap/stdio/putc.c:
```diff
@@ -1,17 +1,10 @@
+	if (f->lock < 0 /*|| !__lockfile(f)*/)
+		return putc_unlocked(c, f);
+	c = putc_unlocked(c, f);
+	//__unlockfile(f);
```
why the commented out code?

elbing replied:
Again old version, perhaps git played with me. Now lock/unlockfile are inocuous functions setting f->lock FILE struct member.

rminnich commented on lib/ap/stdio/ofl.c:
```diff
@@ -0,0 +1,16 @@
+FILE **__ofl_lock()
+{
+	//LOCK(ofl_lock);
+	return &ofl_head;
```
So the lock function doesn't actually lock? So it's not thread safe?

elbing replied:
APE wasn't thread safe. In fact we haven't any posix thread. This is TODO, because that LOCK iis really a pthread_foo at any part...

rminnich commented on lib/ap/stdio/getc.c:
```diff
@@ -1,17 +1,11 @@
+int getc(FILE *f)
+{
+	int c;
+	if (f->lock < 0 /*|| !__lockfile(f)*/)
```
why are the lock/unlock commented out?

elbing replied:
Same, old version...

rminnich commented:
There is some stuff in here that worries me. Overall it's looking great.

rminnich commented:
LGTM

elbing commented:
Fixed some headers more seen during mksh build (still doesn't work).
Removed old bsd.h (remaning from libbsd days). Now ready to merge.
Rest of fixes to get printf family full working in a separate PR, this is now giant to follow it.
